umbral/umbral.py

```diff
@@ -115,98 +116,135 @@ def __bytes__(self):
+                 point_eph_v=None,
+                 bn_sig=None,
+                 e_prime=None,
+                 v_prime=None,
+                 noninteractive_point=None):
+
+        if not point_eph_e and not e_prime:
+            raise ValueError(
+                "Can't make a Capsule from nothing.  Pass either Alice's data (ie, point_eph_e) or Bob's (e_prime). \
+                Passing both is also fine.")
+
+        self._point_eph_e = point_eph_e
+        self._point_eph_v = point_eph_v
+        self._bn_sig = bn_sig
+
+        self._point_eph_e_prime = e_prime
+        self._point_eph_v_prime = v_prime
+        self._point_noninteractive = noninteractive_point
 
         self.cfrags = {}
+        self._contents = None
 
-    @staticmethod
-    def from_bytes(data: bytes, curve: ec.EllipticCurve):
+    @classmethod
+    def from_original_bytes(cls, data: bytes, curve: ec.EllipticCurve):
         """
         Instantiates a Capsule object from the serialized data.
         """
         eph_e = Point.from_bytes(data[0:33], curve)
         eph_v = Point.from_bytes(data[33:66], curve)
         sig = BigNum.from_bytes(data[66:98], curve)
 
-        return Capsule(eph_e, eph_v, sig)
+        return cls(eph_e, eph_v, sig)
+
+    @classmethod
+    def from_reconstructed_bytes(cls, data: bytes, curve: ec.EllipticCurve):
+        """
+        Instantiate a Capsule from serialized data after reconstruction has occurred.
+
+        The most obvious use case is Bob affixing at least m cFrags and then serializing the Capsule.
+        """
+        # TODO: Seems like a job for BytestringSplitter ?
+        e_prime = Point.from_bytes(data[0:33], curve)
+        v_prime = Point.from_bytes(data[33:66], curve)
+        eph_ni = Point.from_bytes(data[66:99], curve)
+
+        return cls(e_prime=e_prime, v_prime=v_prime, noninteractive_point=eph_ni)
+
+    @property
+    def contents(self):
+        return self._contents
 
     def to_bytes(self):
         """
         Serialize the Capsule into a bytestring.
         """
-        eph_e = self.point_eph_e.to_bytes()
-        eph_v = self.point_eph_v.to_bytes()
-        sig = self.bn_sig.to_bytes()
+        eph_e = self._point_eph_e.to_bytes()
+        eph_v = self._point_eph_v.to_bytes()
+        sig = self._bn_sig.to_bytes()
 
         return eph_e + eph_v + sig
 
     def verify(self, params: UmbralParameters):
 
-        e = self.point_eph_e
-        v = self.point_eph_v
-        s = self.bn_sig
+        e = self._point_eph_e
+        v = self._point_eph_v
+        s = self._bn_sig
         h = hash_to_bn([e, v], params)
 
         return params.g * s == v + (e * h)
 
     def attach_cfrag(self, cfrag: CapsuleFrag):
         self.cfrags[cfrag.bn_kfrag_id] = cfrag
 
-    def reconstruct(self):
+    def original_ephemeral_points(self):
+        return self._point_eph_e + self._point_eph_v
+
+    def reconstructed_ephemeral_points(self):
+        return self._point_eph_e_prime + self._point_eph_v_prime
```

Comment:
Please rather call these vars individually for more readability. This also does point addition, did you notice that?

Reply:
...why?

I don't think that three fairly obscure var names are more (or even as) readable as a single method name.

Reply:
...why?

I don't think that three fairly obscure var names are more (or even as) readable as a single method name.

Reply:
...why?

I don't think that three fairly obscure var names are more (or even as) readable as a single method name.

Reply:
Because you're hiding point addition -- this does a few things that we don't want.
    - Since this is a reference implementation, anywhere in the code should be really explicit when we're performing a curve operation. 
    - This is a _dangerous_ operation and shouldn't be exposed as a simple method that the user can confuse for something else. This method is not clear that it performs addition, even with the operator overloading.

Reply:
Another issue here are the names, I just realized:

`original_ephemeral_points` -> It's not returning plural points, after the addition it returns one and they're not the original ones either. Same for `reconstructed_ephemeral_points`.

Reply:
OK, that's a good point.

I still think though that returning them from a single uniform method is much safer.

For the smaller userbase that is willing to assume the liability, the variables are available as private attributes.  But the better approach is to provide a method that does the addition, the correct way, with the correct two values, and to make that method the public default.

So: how about if these two methods instead return a tuple of the points?

This also has the benefit of forcing the candor in the implementation that you're after, ie:

`sum(capsule.original_ephemeral_points) * priv_key`

Reply:
No, that's just clear to me. Can we just leave it as it was? It makes it so much more clear.

Reply:
I think I'm interested in other perspectives here.  @cygnusv, @michwill, @KPrasch?

What's the advantage of *forcing* the user to try to do it properly instead of giving them the method that does (but also providing the primitives)?

Consider:

`shared_key = (capsule.point_eph_e + capsule.point_eph_v) * priv_key`

Have I used the correct two points?
- Did I use the original if that's what I wanted?
- Did I use the reconstructed if that's what I wanted?
- Did I use both from the same set?  I didn't intermix, did I?
- Am I sure I didn't let the `sig` or `noninteractive_point` slip in there?  Definiely `e` and `v`, right?

...especially since there's no great way to `repr` Points from each other in such as way that makes clear their role.

Now, did I add them in the correct order?

Now consider:

`shared_key = sum(capsule.original_ephemeral_points) * priv_key`

There's no way to mess that up.  I know that I have the original points, I know that I have the correct two values, that they're in the right order, that they're added properly.  Nothing is left to chance.

What is bad about this?  Are you suggesting that other libraries refuse to do this on principle?  If so, I want to see the conversation that led to such a design decision - do you have a link?

Reply:
OK, we decided not to do this.  From Slack:

	
jmyles 5:51 PM
Getting back `original_ephemeral_points` for a minute:
5:51
Are you saying that this is never a public notion at all?
tux 5:51 PM
I don’t expect anyone to modify the underlying cryptography of this library. If they do, they probably shouldn’t.
jmyles 5:51 PM
Well, that's not what I asked.  :slightly_smiling_face:
dnunez 5:52 PM
they are in the test because we need to check the serialization
tux 5:52 PM
Right
dnunez 5:52 PM
but a user would never need to inspect those elements
jmyles 5:52 PM
I see
5:52
ok
dnunez 5:52 PM
for a user a capsule is atomic
jmyles 5:52 PM
So they are used by their respective decapsulate functions and never publicly?
5:52
Ahh
5:52
OK
dnunez 5:52 PM
yes
jmyles 5:52 PM
OK yeah, I withdraw then - I misunderstood.
dnunez 5:53 PM
we only use it in umbral, and in the tests of course
5:53
but there shouldn't be used outside that scope
jmyles 5:53 PM
I see.
5:53
Yeah, that completely makes sense now, reading again
dnunez 5:53 PM
in any case, I don't see why is this related with that method that returns the additon of e and v
jmyles 5:54 PM
Well, my drive was to help the user avoid doing that improperly.
5:54
(ie, with the original 'e' and 'v' when they wanted the reconstructed, etc)
5:55
See, this is part of the problem with writing the first time with too few `_private` things :slightly_smiling_face:
dnunez 5:55 PM
ok, that possibly was a problem, yes
tux 5:56 PM
I thought we all knew that all those fragment pieces were parts of the ciphertext?
5:56
intrinsically private sorta :sweat_smile:
dnunez 5:57 PM
it was clear to me, at least, but of course I designed it
5:57
and I have to admit that I'm new to python
tux 5:57 PM
Meh, don’t fret about Python privacy, it’s not real privacy.
dnunez 5:57 PM
the `_private` notation seems weird coming from Java
jmyles 5:57 PM
Yeah, I mean reading again, with the benefit of you're commentary, it makes sense
tux 5:58 PM
It just doesn’t import certain variables prefixed with `_`
dnunez 5:58 PM
since it's not real
jmyles 5:58 PM
But do you see what I was seeing?
dnunez 5:58 PM
not really
5:58
but I understand that there could have been misunderstandings
jmyles 5:58 PM
I saw `shared_key = (capsule.point_eph_e + capsule.point_eph_v) * priv_key` in a public method and thought, "well, if the user wanted to..."
michwill 5:58 PM
Well, you can have two underscores for real :slightly_smiling_face:
tux 5:59 PM
throws papers
jmyles 5:59 PM
hahahaha
5:59
But seriously: the decapsulate methods are good candidates for `__`
5:59
for this reason
dnunez 5:59 PM
I don't see what's the potential problem with the user anyway
tux 6:00 PM
^ Yeah, this is really my main question
6:00
I don’t expect users to go muddling around in the code here.
dnunez 6:00 PM
how could the user screw things there?
tux 6:00 PM
It’s not like they’re gonna go, “Oh, well I need to add to the ciphertext!”
jmyles 6:00 PM
Well, what if they wrote a Bob who decapsulates the capsule in his super special way, but used the original `e` and the reconstrcuted `v`?
tux 6:00 PM
_they should not do that_
dnunez 6:00 PM
basically, it would fail
jmyles 6:01 PM
@tux: if you feel so strongly, then let's have code saying so :slightly_smiling_face:
6:01
@tux: I agree *they should not do that* - that's why I wrote a method that does it the right way for them
6:01
(not yet realizing that they don't need to do it anyway)
dnunez 6:01 PM
but jmyles
6:01
your argument applies to every line of the code
jmyles 6:01 PM
No, only the public interfaces
dnunez 6:02 PM
I dont understand that
jmyles 6:02 PM
If `Capsule.open` calls `__decapsulate_reconstructed`, then it's absolutely obvious that the user isn't allowed to much with the latter without taking complete responsibility for their idiocy
6:02
:slightly_smiling_face:
dnunez 6:02 PM
ok...
jmyles 6:02 PM
I was trying to remove `ReconstructedCapsule` in a way that still forced the user to only be able to do the right thing.
6:03
And, at the time, I thought that I was seeing the user calling `decapsulate` directly, not using the public interfaces on `Capsule` (because yes, I spaced for a second and wasn't thinking of a capsule as an atomic structure)